### PR TITLE
CO-390 automate Codex CLI release detection

### PR DESCRIPTION
**Pull request by Kbediako**
## What
- Add a scheduled and manually dispatchable Codex CLI release detector workflow.
- Add `scripts/codex-cli-release-detector.mjs` plus `npm run codex:release-detect` to compare GitHub `openai/codex` release truth, npm `@openai/codex` dist-tags/time, and current CO policy/pin surfaces.
- Create or reuse one canonical Linear release-intake owner through `create-follow-up --canonical-owner-key`, refresh the selected intake workpad, and emit a machine-readable detection artifact.
- Document operator retry/manual dry-run behavior and cover no-op, new stable, prerelease-only, existing owner, split pins, missing auth, duplicate-prevention, and mismatch paths.

## Why
CO-390 closes the manual gap between upstream Codex CLI releases and the CO-386 release-intake workflow. Detection does not promote versions by itself; it reports represented stable releases, records prerelease-only movement, and fails closed when mutation or release truth is uncertain.

## Validation
- `MCP_RUNNER_TASK_ID=linear-b7074b86-3d38-4dfe-baa9-73b2cc8d686f node scripts/delegation-guard.mjs`
- `node scripts/spec-guard.mjs --dry-run`
- `npm run build`
- `npm run lint` (existing `orchestrator/tests/DelegationMcpHealth.test.ts` warnings only)
- `npm run test` (357 files / 4,954 tests)
- `npm run docs:check`
- `npm run docs:freshness`
- `npm run repo:stewardship`
- `node scripts/diff-budget.mjs`
- `npm run pack:smoke`
- `npm run codex:release-detect -- --dry-run --artifact out/linear-b7074b86-3d38-4dfe-baa9-73b2cc8d686f/codex-cli-release-detection-dry-run.json`

## Review
- Forced standalone review completed with `status=succeeded`, `review_outcome=bounded-success`, and no actionable issues after fixing prior findings.
- Explicit elegance/minimality pass found no safe simplification to apply.

Diff budget override: CO-390 combines the mandatory docs-first packet, upstream release detector implementation, scheduled workflow, and focused regression coverage in one canonical release-intake automation lane; splitting would separate the detector from required acceptance evidence and Linear provenance.

## Summary by CodeRabbit

* **New Features**
  * Automated Codex CLI release detection with scheduled/manual triggers, dry‑run support, artifact output, and canonical intake creation/updating with deduplication; CLI command exposed for detection.

* **Documentation**
  * New PRD, tech spec, action plan, task specs, task checklist and registry/freshness entries detailing scope, gating, validations and handoff requirements.

* **Tests**
  * Comprehensive test suite covering detection scenarios, artifact contents, decision outcomes and mutation behaviours.

* **Chores**
  * CI workflow to run the detector daily/dispatch and upload detection artifacts; package script to run the detector.

### CODE REVIEW

File: scripts/codex-cli-release-detector.mjs [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+  try {
+    releaseTemplate = await readFileImpl(join(repoRoot, RELEASE_INTAKE_TEMPLATE_PATH), 'utf8');
+  } catch {
+    missing.push({ path: RELEASE_INTAKE_TEMPLATE_PATH, error: 'missing release-intake template' });
+  }
+
+  const policy = surfaces.find((surface) => surface.path === 'docs/guides/codex-version-policy.md');
+  const policyContent = policy ? await readFileImpl(join(repoRoot, policy.path), 'utf8') : '';
+  const policyVersions = extractVersions(policyContent).filter((version) => !parseSemver(version)?.prerelease);
+  const lastAuditedVersion = sortVersionsDescending(policyVersions.map((version) => ({ version })))[0]?.version ?? null;
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs:339]: Fixed in ac5b68fec: `last_audited_version` now derives from explicit audit evidence via `extractAuditedVersions(...)`, not the max stable version mentioned anywhere in the policy file, and focused coverage exercises the candidate-versus-audited case.

File: scripts/codex-cli-release-detector.mjs [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+} = {}) {
+  if (typeof fetchImpl !== 'function') {
+    throw new Error('No fetch implementation available for upstream release detection.');
+  }
+  const githubHeaders = {
+    Accept: 'application/vnd.github+json',
+    'X-GitHub-Api-Version': '2022-11-28',
+    ...(githubToken ? { Authorization: `Bearer ${githubToken}` } : {})
+  };
+  const githubUrl = `https://api.github.com/repos/${githubRepo}/releases?per_page=20`;
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs]: Fixed in ac5b68fec: GitHub release fetch now requests `per_page=100` and paginates bounded pages before classifying stable/prerelease releases, with focused pagination coverage.

File: docs/PRD-linear-b7074b86-3d38-4dfe-baa9-73b2cc8d686f.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+| Upstream release truth | Manual posture lanes have captured release/npm facts after a human starts the audit. | GitHub release truth and npm `@openai/codex` dist-tags/time together establish whether a canonical upstream release candidate exists. | A detector compares both sources and triggers one canonical Linear intake issue only when the release is new or materially untracked. |
+| Linear intake | Release lanes can be created manually from operator context. | CO-386 release-intake template defines the canonical issue shape for release-intake work. | CO-390 funnels release discovery into one canonical Linear intake issue and updates/deduplicates instead of spraying follow-ups. |
+
+Explicitly out-of-scope differences: direct Codex CLI promotion, workflow pin changes, cloud-canary execution, pack-smoke rebaseline, marketplace compatibility redesign, implementation file edits, and Linear mutation from this child lane.
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply docs/PRD-linear-b7074b86-3d38-4dfe-baa9-73b2cc8d686f.md:60]: @coderabbitai Fixed in ac5b68fec: the parity matrix now has blank lines before and after the markdown table, and `npm run docs:check` passed locally after the fix.

**Comment by Kbediako** [comment]
Parent orchestration gate: before ready-review/merge, please add explicit coverage that a required test pin surface (at minimum `tests/pack-smoke.spec.ts`) fails closed when it contains an unversioned `@openai/codex` install. The current `e197d50ae` fix appears to cover the generic required-surface loop, but the focused regression only mutates `.github/workflows/core-lane.yml`; validator consensus is to prove both workflow and test pin surfaces for the latest P2 review finding.

File: scripts/codex-cli-release-detector.mjs [hunk 4]
```diff
@@ -0,0 +1,875 @@
...
+  };
+  const remainingRaw = get('x-ratelimit-remaining');
+  const limitRaw = get('x-ratelimit-limit');
+  const resetRaw = get('x-ratelimit-reset');
+  return {
+    remaining: remainingRaw === null ? null : Number(remainingRaw),
+    limit: limitRaw === null ? null : Number(limitRaw),
+    reset_epoch_seconds: resetRaw === null ? null : Number(resetRaw),
+    uncertain: remainingRaw === null || Number.isNaN(Number(remainingRaw)) || Number(remainingRaw) <= 0
+  };
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs:167]: @coderabbitai Fixed in 56c0678ea: zero remaining quota is accepted when the response completed, while missing or non-numeric quota still fails closed. Focused coverage verifies `remaining: 0` with `uncertain: false`.

File: scripts/codex-cli-release-detector.mjs [hunk 5]
```diff
@@ -0,0 +1,870 @@
...
+    reset_epoch_seconds: resetValues.length > 0 ? Math.max(...resetValues) : null,
+    uncertain: rateLimits.some((limit) => limit.uncertain)
+  };
+}
+
+async function fetchGithubReleases(fetchImpl, githubRepo, headers) {
+  const releases = [];
+  const rateLimits = [];
+  const perPage = 100;
+  const maxPages = 10;
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs]: Fixed in 956a20e2a: GitHub release discovery no longer has a page-count cap; it requests 100 releases per page and continues until the API returns a short page. Added regression coverage for a stable release after more than ten prerelease pages.

File: scripts/codex-cli-release-detector.mjs [hunk 6]
```diff
@@ -0,0 +1,870 @@
...
+  const match = content.match(pattern);
+  return match?.[1] ?? null;
+}
+
+function extractAuditedVersions(content) {
+  const versions = new Set();
+  const patterns = [
+    /\baudited(?: official)? `rust-v(\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?)`/gi,
+    /\baudited[^.\n]*npm `@openai\/codex@(\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?)`/gi,
+    /\baudited[^.\n]*Codex CLI `(\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?)`/gi
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs]: Fixed in 956a20e2a: audited versions now come only from explicit release-intake evidence markers (`audited rust-v...` or audited npm `@openai/codex@...`), excluding control-seam-only `audited Codex CLI ...` lines. Added regression coverage for that case.

File: scripts/codex-cli-release-detector.mjs [hunk 7]
```diff
@@ -0,0 +1,870 @@
...
+    return headers[name] ?? headers[name.toLowerCase()] ?? null;
+  };
+  const remainingRaw = get('x-ratelimit-remaining');
+  const limitRaw = get('x-ratelimit-limit');
+  const resetRaw = get('x-ratelimit-reset');
+  return {
+    remaining: remainingRaw === null ? null : Number(remainingRaw),
+    limit: limitRaw === null ? null : Number(limitRaw),
+    reset_epoch_seconds: resetRaw === null ? null : Number(resetRaw),
+    uncertain: remainingRaw === null || Number.isNaN(Number(remainingRaw)) || Number(remainingRaw) <= 0
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs]: Fixed in 56c0678ea: zero remaining quota on a complete response is not treated as uncertain; missing or non-numeric quota remains fail-closed. Added regression coverage for `remaining: 0, uncertain: false`.

File: scripts/codex-cli-release-detector.mjs [hunk 8]
```diff
@@ -0,0 +1,875 @@
...
+  return match?.[1] ?? null;
+}
+
+export function classifyGithubReleases(releases) {
+  const normalized = [];
+  for (const release of releases ?? []) {
+    const tagName = release.tag_name ?? release.tagName;
+    const version = releaseVersionFromTag(tagName);
+    const parsed = parseSemver(version);
+    if (!parsed) {
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs:91]: Fixed in e37e6255e: `classifyGithubReleases` skips `draft`/`isDraft` releases before stable/prerelease classification, with regression coverage for draft stable releases.

File: tests/codex-cli-release-detector.spec.ts [hunk 9]
```diff
@@ -0,0 +1,843 @@
...
+      artifactPath: 'out/detection.json',
+      fetchImpl: mockFetch({ stable: '0.126.0' }),
+      dryRun: true,
+      env: {}
+    });
+
+    expect(exitCode).toBe(0);
+    expect(artifact.decision_state).toBe('new_audit_required');
+    expect(artifact.last_audited_version).toBe('0.125.0');
+    expect(artifact.mutation_result.action).toBe('dry_run');
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply tests/codex-cli-release-detector.spec.ts]: @coderabbitai Fixed in e9670ebc6: the tests now assert `artifact.current_co.policy.last_audited_version` for the detector contract, including the written artifact schema check. Focused detector suite is 24/24.

File: scripts/codex-cli-release-detector.mjs [hunk 10]
```diff
@@ -0,0 +1,875 @@
...
+      ].map((match) => match[1]);
+      surfaceContents.set(surface, content);
+      surfaces.push({
+        path: surface,
+        versions: extractVersions(content),
+        install_pins: installPins
+      });
+      if (VERSIONED_PIN_REQUIRED_SURFACES.has(surface) && installPins.length === 0) {
+        missing.push({ path: surface, error: 'missing versioned @openai/codex install pin' });
+      }
```
[the remark this answers is not in the document]
Comment (Kbediako) [reply scripts/codex-cli-release-detector.mjs]: Fixed in e37e6255e and extended in 6bb923d11: required pin surfaces now fail closed on any unversioned `@openai/codex` install, including `npm i -g`, even when a versioned install also exists. Focused coverage includes bare install and install-alias paths.

**Comment by Kbediako** [comment]
Addressed the explicit test-coverage gate in e37e6255e, 6bb923d11, and e9670ebc6: `tests/pack-smoke.spec.ts` required-pin coverage now fails closed for both missing versioned pins and versioned-plus-bare unversioned `@openai/codex` installs, including `npm i -g` aliases. Focused detector suite is 24/24, and the final dry-run artifact reports `prerelease_observed`.

**Comment by Kbediako** [comment]
Latest status: PR #702 is pushed at e9670ebc6 with unresolved review threads at 0, CodeRabbit SUCCESS, and Cloud Canary SUCCESS. Core Lane is red only at Spec guard due pre-existing Mar 27 stale specs (`linear-488135bf`, `linear-964ca955`, `linear-da290822`, `linear-e913b2ab`), matching the local docs-freshness stale cohort with `blocking_changed_paths=[]`. I filed that out-of-scope baseline blocker as CO-401 and moved CO-390 to Blocked rather than handing this PR to In Review while required checks are red.
